src/System.Runtime.Numerics/src/System/Numerics/Complex.cs

```diff
@@ -295,7 +301,90 @@ public static Complex Exp(Complex value)
+                // Not only is this more expensive than necessary, it also fails to preserve certain expected
+                // symmetries, such as that the square root of a pure negative is a pure imaginary, and that the
+                // square root of a pure imaginary has exactly equal real and imaginary parts. This all goes
+                // back to the fact that Math.PI is not stored with infinite precision, so taking half of Math.PI
+                // does not land us on an argument with cosine exactly equal to zero.
+
+                // To find a fast and symmetry-respecting formula for complex square root,
+                // note x + i y = \sqrt{a + i b} implies x^2 + 2 i x y - y^2 = a + i b,
+                // so x^2 - y^2 = a and 2 x y = b. Cross-substitute and use the quadratic formula to obtain
+                //   x = \sqrt{\frac{\sqrt{a^2 + b^2} + a}{2}}  y = \pm \sqrt{\frac{\sqrt{a^2 + b^2} - a}{2}}
+                // There is just one complication: depending on the sign on a, either x or y suffers from
+                // cancelation when |b| << |a|. We can get aroud this by noting that our formulas imply
+                // x^2 y^2 = b^2 / 4, so |x| |y| = |b| / 2. So after computing the one that doesn't suffer
+                // from cancelation, we can compute the other with just a division. This is basically just
+                // the right way to evaluate the quadratic formula without cancelation.
+
+                // All this reduces our total cost to two sqrts and a few flops, and it respects the desired
+                // symmetries. Much better than atan + cos + sin!
+
+                // The signs are a matter of choice of branch cut, which is traditionally taken so x > 0 and sign(y) = sign(b).
+      
+                // If the components are too large, Hypot will overflow, even though the subsequent sqrt would
+                // make the result representable. To avoid this, we re-scale (by exact powers of 2 for accuracy)
+                // when we encounter very large components to avoid intermediate infinities.
+                bool rescale = false;
+                const double max = Double.MaxValue / 2.0;
+                if ((Math.Abs(value._real) >= max) || (Math.Abs(value._imaginary) >= max))
+                {
+                    value._real *= 0.25;
+                    value._imaginary *= 0.25;
+                    rescale = true;
+                }
+                // 1. Note that, if one component is very large and the other is very small, this re-scale could cause
+                // the very small component to underflow. This is not a problem, though, because the neglected term
+                // is ~ (very small) / (very large), which would underflow anyway.
+                // 2. Note that our test for re-scaling requires two abs and two floating point comparisons.
+                // For the typical non-overflowng case, it might be faster to just test for IsInfinity after Hypot,
```

Comment:
I'd prefer to just mention this stuff in the commit description, or in the PR description / comments, rather than leave it in the source code forever. If it's not actionable then we probably don't need to capture it here.